ports/esp32/machine_timer.c

```diff
@@ -60,74 +61,122 @@ uint32_t machine_timer_freq_hz(void) {
     return freq / TIMER_DIVIDER;
 }
 
+// Called by board port soft reset routine to deactivate all timers before reboot.
 void machine_timer_deinit_all(void) {
-    // Disable, deallocate and remove all timers from list
+    // Disable all timers in list, gc will handle deletion of allocated python memory
     machine_timer_obj_t **t = &MP_STATE_PORT(machine_timer_obj_head);
     while (*t != NULL) {
-        machine_timer_deinit(*t);
         machine_timer_obj_t *next = (*t)->next;
-        m_del_obj(machine_timer_obj_t, *t);
+        machine_timer_deinit(*t);
         *t = next;
     }
 }
 
 static void machine_timer_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
     machine_timer_obj_t *self = self_in;
     qstr mode = self->repeat ? MP_QSTR_PERIODIC : MP_QSTR_ONE_SHOT;
-    uint64_t period = self->period / (machine_timer_freq_hz() / 1000); // convert to ms
-    #if SOC_TIMER_GROUP_TIMERS_PER_GROUP == 1
-    mp_printf(print, "Timer(%u, mode=%q, period=%lu)", self->group, mode, period);
-    #else
-    mp_printf(print, "Timer(%u, mode=%q, period=%lu)", (self->group << 1) | self->index, mode, period);
-    #endif
-}
+    uint64_t period = 0;
+    mp_uint_t timer_id = -1;
 
-machine_timer_obj_t *machine_timer_create(mp_uint_t timer) {
+    if (self->is_virtual) {
+        // Virtual timer
+        timer_id = -1;
 
-    machine_timer_obj_t *self = NULL;
-    #if SOC_TIMER_GROUP_TIMERS_PER_GROUP == 1
-    mp_uint_t group = timer & 1;
-    mp_uint_t index = 0;
-    #else
-    mp_uint_t group = (timer >> 1) & 1;
-    mp_uint_t index = timer & 1;
-    #endif
+        // convert from us to ms
+        period = self->period / 1000;
+    } else {
+        // Hardware timer
+        #if SOC_TIMER_GROUP_TIMERS_PER_GROUP == 1
+        timer_id = self->group;
+        #else
+        timer_id = (self->group << 1) | self->index;
+        #endif
 
-    // Check whether the timer is already initialized, if so use it
-    for (machine_timer_obj_t *t = MP_STATE_PORT(machine_timer_obj_head); t; t = t->next) {
-        if (t->group == group && t->index == index) {
-            self = t;
-            break;
-        }
+        // convert from machine timer value to ms
+        period = self->period / (machine_timer_freq_hz() / 1000);
     }
-    // The timer does not exist, create it.
-    if (self == NULL) {
-        self = mp_obj_malloc(machine_timer_obj_t, &machine_timer_type);
-        self->group = group;
-        self->index = index;
-        self->handle = NULL;
+
+    mp_printf(print, "Timer(%d, mode=%q, period=%lu)", timer_id, mode, (mp_int_t)period);
+}
+
+machine_timer_obj_t *machine_timer_create(mp_int_t timer_id) {
+    machine_timer_obj_t *self = NULL;
+
+    if (timer_id == -1) {
+        // Virtual timer creation
+        self = mp_obj_malloc_with_finaliser(machine_timer_obj_t, &machine_timer_type);
```

Comment:
Given that this module now depends on finalisers to prevent a GC from breaking entries in our linked list, should we check for `MICROPY_ENABLE_FINALISER`? Some other modules have a explicit compile-time check at the top of the file, something like:

```c
#if !MICROPY_ENABLE_FINALISER
#error "machine.Timer requires MICROPY_ENABLE_FINALISER."
#endif
```